src/hotspot/share/gc/parallel/psPromotionManager.inline.hpp

```diff
@@ -126,173 +127,187 @@ inline void PSPromotionManager::push_contents(oop obj) {
-          CollectedHeap::fill_with_object(cast_from_oop<HeapWord*>(new_obj), new_obj_size);
-        }
-      } else if (!_young_lab.unallocate_object(cast_from_oop<HeapWord*>(new_obj), new_obj_size)) {
+    log_develop_trace(gc, scavenge)("{%s %s " PTR_FORMAT " -> " PTR_FORMAT " (%d)}",
+                                    new_obj_is_tenured ? "copying" : "tenuring",
+                                    new_obj->klass()->internal_name(),
+                                    p2i((void *)o), p2i((void *)new_obj), new_obj->size());
+
+    // Do the size comparison first with new_obj_size, which we
+    // already have. Hopefully, only a few objects are larger than
+    // _min_array_size_for_chunking, and most of them will be arrays.
+    // So, the is->objArray() test would be very infrequent.
+    if (new_obj_size > _min_array_size_for_chunking &&
+        new_obj->is_objArray() &&
+        PSChunkLargeArrays) {
+      // we'll chunk it
+      push_depth(ScannerTask(PartialArrayScanTask(o)));
+      TASKQUEUE_STATS_ONLY(++_arrays_chunked; ++_array_chunk_pushes);
+    } else {
+      // we'll just push its contents
+      push_contents(new_obj);
+    }
+  } else {
+    // We lost, someone else "owns" this object.
+    // Ensure loads of or from the forwardee follow all changes that
+    // preceeded the release-cmpxchg that performed the forwarding in
+    // another thread.  Don't let them float above the failed cmpxchg by
+    // this thread, which is effectively a relaxed-load.
+    OrderAccess::acquire();
+
+    assert(o->is_forwarded(), "Object must be forwarded if the cas failed.");
+
+    // Try to deallocate the space.  If it was directly allocated we cannot
+    // deallocate it, so we have to test.  If the deallocation fails,
+    // overwrite with a filler object.
+    if (new_obj_is_tenured) {
+      if (!_old_lab.unallocate_object(cast_from_oop<HeapWord*>(new_obj), new_obj_size)) {
         CollectedHeap::fill_with_object(cast_from_oop<HeapWord*>(new_obj), new_obj_size);
       }
-
-      // don't update this before the unallocation!
-      // Using acquire though consume would be accurate for accessing new_obj.
-      new_obj = o->forwardee_acquire();
+    } else if (!_young_lab.unallocate_object(cast_from_oop<HeapWord*>(new_obj), new_obj_size)) {
+      CollectedHeap::fill_with_object(cast_from_oop<HeapWord*>(new_obj), new_obj_size);
     }
-  } else {
-    assert(o->is_forwarded(), "Sanity");
-    new_obj = o->forwardee_acquire();
-  }
 
-  // This code must come after the CAS test, or it will print incorrect
-  // information.
-  log_develop_trace(gc, scavenge)("{%s %s " PTR_FORMAT " -> " PTR_FORMAT " (%d)}",
-                                  should_scavenge(&new_obj) ? "copying" : "tenuring",
-                                  new_obj->klass()->internal_name(), p2i((void *)o), p2i((void *)new_obj), new_obj->size());
+    // don't update this before the unallocation!
```

Comment:
This comment gives me the impression that there is some racy going on here, and deallocation and update to `new_obj` must be ordered this way. However, the actual reason is that  the old value of`new_obj` is used for deallocation. I think it's best to remove this comment; it doesn't really say anything interesting.

Reply:
That comment went away as part of latest change to avoid reloading the forwardee.